### PR TITLE
[webkit.RefCntblBaseVirtualDtor] Allow CRTP classes without a virtual destructor.

### PR DESCRIPTION
**Pull request by rniwa**
Exempt CRTP (Curiously Recurring Template Pattern) classes with a delete operation acting on "this" pointer with an appropriate cast from the requirement that a ref-countable superclass must have a virtual destructor.

To do this, this PR introduces new DerefAnalysisVisitor, which looks for a delete operation with an explicit cast to the derived class in a base class.

This PR also changes the checker so that we only check a given class's immediate base class instead of all ancestor base classes in the class hierarchy. This is sufficient because the checker will eventually see the definition for every class in the class hierarchy and transitively proves every ref-counted base class has a virtual destructor or deref function which casts this pointer back to the derived class before deleting. Without this change, we would keep traversing the same list of base classes whenever we encounter a new subclass, which is wholly unnecessary.

It's possible for DerefAnalysisVisitor to come to a conclusoin that there isn't enough information to determine whether a given templated superclass invokes delete operation on a subclass when the template isn't fully specialized for the subclass. In this case, we return std::nullopt in HasSpecializedDelete, and visitCXXRecordDecl will skip this declaration. This is okay because the checker will eventually see a concreate fully specialized class definition if it ever gets instantiated.

### CODE REVIEW

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp [hunk 1]
```diff
@@ -51,92 +169,137 @@ class RefCntblBaseVirtualDtorChecker
...
+          if (T.isNull())
+            continue;
+
+          const CXXRecordDecl *C = T->getAsCXXRecordDecl();
+          if (!C)
+            continue;
+
+          if (auto *CTSD = dyn_cast<ClassTemplateSpecializationDecl>(C)) {
+            auto &Args = CTSD->getTemplateArgs();
+            for (unsigned i = 0; i < Args.size(); ++i) {
```
Comment (haoNoQ) [inline clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:193]: If you like you can do a `CTSD->getTemplateArgs().asArray()` and do a range-based loop. It's going to be an `llvm::ArrayRef` so you can store it by value too.

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp [hunk 2]
```diff
@@ -11,16 +11,134 @@
...
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/SetVector.h"
 #include <optional>
 
 using namespace clang;
 using namespace ento;
 
 namespace {
+
+class DerefAnalysisVisitor
```
Comment (haoNoQ) [inline clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:28]: 🚲🏠: by "deref" you mean "decrement reference count" right? Because when I see it in the compiler I mostly think of operations `*` and `->` and `[]` and such. This is readable when it's next to a similarly named thing named "ref" but in this case I was confused for a few minutes, like what does this have to do with templates?

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp [hunk 3]
```diff
@@ -11,16 +11,134 @@
...
+  }
+
+  bool VisitCallExpr(const CallExpr *CE) {
+    auto *Callee = CE->getCallee();
+    while (auto *Expr = dyn_cast<CastExpr>(Callee))
+      Callee = Expr->getSubExpr();
+    if (auto *DeclRef = dyn_cast<DeclRefExpr>(Callee)) {
+      auto *Decl = DeclRef->getDecl();
+      if (auto *VD = dyn_cast<VarDecl>(Decl)) {
+        if (auto *Init = VD->getInit()) {
```
[the remark this answers is not in the document]
Comment (haoNoQ) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:72]: Never mind, you cannot assign lambdas this way.

This means that your `getInit()` branch is probably unnecessary; the compiler knows at compile time which `operator()()` is called(?)

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp (hunk 2 above)
Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:28]: Oh, this is `deref` function (i.e. function by the name of `deref`) analysis visitor.
Maybe I should call it `DerefFuncVisitor` instead?

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp (hunk 1 above)
Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:193]: oh, neat. I'll go do that.

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp [hunk 4]
```diff
@@ -11,16 +11,134 @@
...
+    }
+    return false;
+  }
+
+  bool VisitBody(const Stmt *Body) {
+    if (!Body)
+      return false;
+
+    auto [It, IsNew] = VisitedBody.insert(Body);
+    if (!IsNew) // This body is recursive
```
[the remark this answers is not in the document]
Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:44]: We could do that.

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp [hunk 5]
```diff
@@ -11,16 +11,134 @@
...
+  std::optional<bool> HasSpecializedDelete(CXXMethodDecl *Decl) {
+    if (auto *Body = Decl->getBody())
+      return VisitBody(Body);
+    if (auto *Tmpl = Decl->getTemplateInstantiationPattern())
+      return std::nullopt; // Indeterminate. There was no concrete instance.
+    return false;
+  }
+
+  bool VisitCallExpr(const CallExpr *CE) {
+    auto *Callee = CE->getCallee();
```
[the remark this answers is not in the document]
Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:66]: Sure, we can do that.

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp (hunk 2 above)
Comment (haoNoQ) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:28]: I'd even go with like `UnsafeDeleteVisitor` or something. Because fundamentally you're looking for `delete`, and `deref` is just one of the forms it takes, or something 😅

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:72]: What do you mean by `getInit()` branch is unnecessary?

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp [hunk 6]
```diff
@@ -11,16 +11,134 @@
...
+          if (auto *Lambda = dyn_cast<LambdaExpr>(Init))
+            return VisitBody(Lambda->getBody());
+        }
+      } else if (auto *FD = dyn_cast<FunctionDecl>(Decl))
+        return VisitBody(FD->getBody());
+    }
+    return false;
+  }
+
+  bool VisitCXXMemberCallExpr(const CXXMemberCallExpr *MCE) {
```
[the remark this answers is not in the document]
Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:82]: Sure, we can do that.

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (haoNoQ) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:72]: I mean, the only thing you can call on a lambda-type variable, is that lambda's `operator()()`. And it can't be any other lambda's `operator()()` because you can't assign lambdas to each other even when they have the same signature. So I suspect that if you simply ask the AST for `CE->getCalleeDecl()` or `CE->getDirectCallee()` or something of that nature, it'll give you the declaration of `operator()()` without any extra steps, and it'll fit in naturally with the rest of the code so the `getInit()` branch can be completely deleted.

I might be wrong, needs testing.

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp (hunk 2 above)
Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:28]: Hm... but it's important that the function in question is `deref()` though. In that, that's what guarantees the object to be eventually deleted appropriately. Maybe `DerefFuncDeleteExprVisitor`?

Comment (haoNoQ) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:28]: Sounds great to me!

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (haoNoQ) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:72]: In other words, it's a type-system problem. You're not supposed to be emulating data flow analysis in order to solve it.

If it was a raw function pointer then yeah, you'd have to look at the initializer.

But if it's a lambda then all you need should be in the type of the variable. You don't need to jump from statement to statement (from lambda invocation expression to lambda variable initializer expression) in order to figure it out.

It's fairly unobvious because it may look like the type of the variable is `auto` which doesn't really say much. But in the AST all the `auto`s are fully resolved, and in this case this `auto` is resolved to the lambda's underlying anonymous class type, from which the `operator()()` can be trivially looked up.

[the remark this answers is not in the document]
Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:72]: Oh I see. Indeed, it looks like we can get away with just:
```cpp
  bool VisitCallExpr(const CallExpr *CE) {
    const Decl *D = CE->getCalleeDecl();
    if (D && D->hasBody())
      return VisitBody(D->getBody());
    return false;
  }
```

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:44]: Actually, the only way we'd hit this code path is if the function is recursive and we're still in the middle of recursively visiting the function definition so the best we can do here is to return false (since we haven't finished traversing the function body yet).

[the remark this answers is not in the document]
Comment (haoNoQ) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:44]: Oh ok it's a one-shot visitor that scans one function and dies. But if it is instantiated again to scan another function that calls a previously visited function, it'll re-scan it from scratch.

So like, this probably ensures termination but there's also probably room for major optimizations there. We don't really have strict performance requirements, esp. for project-specific checks, but at a glance this could matter on some TUs, so it may be a good idea to keep an eye on performance.

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp [hunk 7]
```diff
@@ -11,16 +11,118 @@
...
+          auto PointeeType = PtrType->getPointeeType();
+          while (auto *ET = dyn_cast<ElaboratedType>(PointeeType)) {
+            if (ET->isSugared())
+              PointeeType = ET->desugar();
+          }
+          if (auto *ParmType = dyn_cast<TemplateTypeParmType>(PointeeType)) {
+            if (ArgList) {
+              auto ParmIndex = ParmType->getIndex();
+              auto Type = ArgList->get(ParmIndex).getAsType();
+              if (auto *RD = dyn_cast<RecordType>(Type)) {
```
Comment (haoNoQ) [inline clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:91]: `Type->getAsCXXRecordDecl() == ClassDecl` is a bit shorter, and you probably don't even need to null-check, because `ClassDecl` is never null anyway.

**Review by haoNoQ — APPROVED** [review]
LGTM now!

I have one style comment but that's it, everything looks good.

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:44]: Yeah. We could keep the visitor around for the duration of the checker & cache the results. Will address that in a follow up.

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp [hunk 8]
```diff
@@ -11,16 +11,116 @@
...
+                             const CXXRecordDecl *ClassDecl)
+      : ArgList(&ArgList), ClassDecl(ClassDecl) {}
+
+  DerefFuncDeleteExprVisitor(const CXXRecordDecl *ClassDecl)
+      : ClassDecl(ClassDecl) {}
+
+  std::optional<bool> HasSpecializedDelete(CXXMethodDecl *Decl) {
+    if (auto *Body = Decl->getBody())
+      return VisitBody(Body);
+    if (auto *Tmpl = Decl->getTemplateInstantiationPattern())
```
Comment (chapuni) [inline clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:61]: This causes a warning.

```
clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:61:15: warning: variable 'Tmpl' set but not used [-Wunused-but-set-variable]
```

Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:61]: Oh oops, sorry about that.

Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:61]: Here's a PR to fix it: https://github.com/llvm/llvm-project/pull/93403
